Using Blocks API to handle ECE cart requests

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This PR expands the usage of the Blocks API (Cart API) to requests related to the cart itself: when adding items to the cart, retrieving items, removing, emptying, etc.

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

- Checkout and build this branch on your test environment (`tweak/using-blocks-api-to-process-cart-requests`)
- Connect your Stripe account
- Enable express payment methods (Google Pay, Apple Pay, Link)
- As a shopper, purchase any products using the multiple ways available: product page, cart page (shortcode and block), checkout (shortcode and block), pay for order.
- Confirm all flows are still working correctly

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Using Blocks API to handle ECE cart requests

wjrosa committed:
Adding compatibility classes

wjrosa committed:
Adding specific unit test

wjrosa committed:
Adding specific unit test

wjrosa committed:
Installing HTML entities npm package

wjrosa committed:
Adding new transformer class

wjrosa committed:
Fix cart parsing by using the new transformer class

wjrosa committed:
Fix debounce call

wjrosa committed:
Fix filter name

wjrosa committed:
Removing newly added actions to add them later

wjrosa committed:
Changelog and reame entries

wjrosa committed:
Fix display items in product page

wjrosa committed:
Fix add to cart

wjrosa committed:
Adding new method to handle display items

Commit:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

annemirasol reviewed (changes requested):
Great work! Found a shipping rates-related bug in shortcode cart/checkout:

1. Clear your cookies.
2. A a guest shopper, add a shippable product to cart.
3. Go to shortcode cart or shortcode checkout, and click Google Pay.
4. You will see this console error: 

```
v3/?ver=3.0:1 Uncaught (in promise) IntegrationError: When `shippingAddressRequired` is true, you must specify `shippingRates`.
```

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa committed:
Fix anonymous shipping + add discount support

wjrosa commented:
> Great work! Found a shipping rates-related bug in shortcode cart/checkout:
> 
> 1. Clear your cookies.
> 2. A a guest shopper, add a shippable product to cart.
> 3. Go to shortcode cart or shortcode checkout, and click Google Pay.
> 4. You will see this console error:
> 
> ```
> v3/?ver=3.0:1 Uncaught (in promise) IntegrationError: When `shippingAddressRequired` is true, you must specify `shippingRates`.
> ```

Good catch, Anne! I fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3780/commits/2af80a8d3d4e650763cbe06e0a3155ab5c3881d2 . Also added support for discounts.

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa committed:
Fix discount line item

annemirasol commented:
> Good catch, Anne! I fixed it in [2af80a8](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/2af80a8d3d4e650763cbe06e0a3155ab5c3881d2) . Also added support for discounts.

@wjrosa Weird, I am still getting the same error. I pulled the latest commits and re-built the source.

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa commented:
> > Good catch, Anne! I fixed it in [2af80a8](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/2af80a8d3d4e650763cbe06e0a3155ab5c3881d2) . Also added support for discounts.
> 
> @wjrosa Weird, I am still getting the same error. I pulled the latest commits and re-built the source.

Yeah, that's strange. I just tested again using the checkout, product, and cart pages. Anonymous and when logged in 🤔 . Did you try anything different from the steps you shared above?

annemirasol commented:
> > > Good catch, Anne! I fixed it in [2af80a8](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/2af80a8d3d4e650763cbe06e0a3155ab5c3881d2) . Also added support for discounts.
> > 
> > 
> > @wjrosa Weird, I am still getting the same error. I pulled the latest commits and re-built the source.
> 
> Yeah, that's strange. I just tested again using the checkout, product, and cart pages. Anonymous and when logged in 🤔 . Did you try anything different from the steps you shared above?

Same steps.

Do you have shipping zones defined?  I have this:

<img width="400" alt="Screenshot 2025-01-30 at 10 06 43 AM" src="https://github.com/user-attachments/assets/db1f157a-0325-48f3-9d86-1463119fcb5d" />

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa committed:
Fix free shipping

wjrosa commented:
> f4c102a

Right! The issue was specific to free shipping. I was able to reproduce it using your setup and fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3780/commits/f4c102a7313f1dba94a7b90ee6f63e3e3ab31b1f

wjrosa committed:
Better fix

annemirasol commented:
> > [f4c102a](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/f4c102a7313f1dba94a7b90ee6f63e3e3ab31b1f)
> 
> Right! The issue was specific to free shipping. I was able to reproduce it using your setup and fixed it in [f4c102a](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/f4c102a7313f1dba94a7b90ee6f63e3e3ab31b1f)

Fix verified! Thanks!


Found a different issue, this time with variable products.

1. Go to the product page of a variable product, e.g. https://my-test-store.com/product/hoodie/
2. Select a variation, e.g. for Hoodie, "Yes" or "No" for the Logo variation.
3. Click Google Pay.
4. See console errors, and notice how the Google Pay modal no longer lets you choose any address.


Note that there is a PHP notice seen when loading a variable product page. I've filed the issue in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3815. This notice is also present in `develop` and does not seem to be affecting ability to purchase the item via express payment.

Edited to include screenshot of console errors:
<img width="281" alt="Screenshot 2025-01-31 at 10 27 45 AM" src="https://github.com/user-attachments/assets/3d2d15f2-0b5b-4f21-8e11-0265e47983d2" />

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa committed:
Fix product variations

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa commented:
> > > [f4c102a](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/f4c102a7313f1dba94a7b90ee6f63e3e3ab31b1f)
> > 
> > 
> > Right! The issue was specific to free shipping. I was able to reproduce it using your setup and fixed it in [f4c102a](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/f4c102a7313f1dba94a7b90ee6f63e3e3ab31b1f)
> 
> Fix verified! Thanks!
> 
> Found a different issue, this time with variable products.
> 
> 1. Go to the product page of a variable product, e.g. https://my-test-store.com/product/hoodie/
> 2. Select a variation, e.g. for Hoodie, "Yes" or "No" for the Logo variation.
> 3. Click Google Pay.
> 4. See console errors, and notice how the Google Pay modal no longer lets you choose any address.
> 
> Note that there is a PHP notice seen when loading a variable product page. I've filed the issue in #3815. This notice is also present in `develop` and does not seem to be affecting ability to purchase the item via express payment.
> 
> Edited to include screenshot of console errors: <img alt="Screenshot 2025-01-31 at 10 27 45 AM" width="281" src="https://private-user-images.githubusercontent.com/730823/408647246-3d2d15f2-0b5b-4f21-8e11-0265e47983d2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg2MTYyMTEsIm5iZiI6MTczODYxNTkxMSwicGF0aCI6Ii83MzA4MjMvNDA4NjQ3MjQ2LTNkMmQxNWYyLTBiNWItNGYyMS04ZTExLTAyNjVlNDc5ODNkMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIwM1QyMDUxNTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMTA3YjExYzNiZTRlZjMyZTU2ZDg0NTgyYTcwNjNmMjllMWJmYTllZDlmZGQ0YmU3ODQzNmRkNGM4NjE0YWY1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pn4p6sVbbxg192-MupVMj-QsL4MlracF1U3U5kbuJXc">

Another great catch, Anne! 👍  Fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3780/commits/57a8a79a96e85707f16b3f59a7378e0d19736497. But it will just work with WC 9.7.0 since this part of the code was fixed [here](https://github.com/woocommerce/woocommerce/pull/53458). So, upon merging this PR we may need to update the min. required WC version 🤔  (after waiting for the release).

wjrosa committed:
Merge branch 'develop' into tweak/using-blocks-api-to-process-cart-requests

wjrosa committed:
Supporting legacy endpoints when variations are present

wjrosa commented:
@annemirasol to avoid blocking this PR, I decided to support the legacy version when variations are present in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3780/commits/e123a0c91a78135afbe934e0eab4349ef24df421. Let me know what you think 👀

annemirasol reviewed (approved):
> to avoid blocking this PR, I decided to support the legacy version when variations are present in https://github.com/woocommerce/woocommerce-gateway-stripe/commit/e123a0c91a78135afbe934e0eab4349ef24df421. Let me know what you think 👀

Makes sense to me!

LGTM, no regression detected -- can complete purchases across different pages and using different payment methods! 🚢
